### PR TITLE
Rate user feature

### PR DESCRIPTION
**Pull request by arkanoider**
Hi @grunch ,

this is related to :

https://github.com/MostroP2P/mostro-core/pull/73

i tested that now new users are added correctly when they send a new message to `mostrod`, need to check more what happens with latest trade index. Now users have `identity key` ad primary key in db, i removed crud from users in mostro-core which was limiting the use of `string`.

You will see `cargo.toml` pointing to `mostro-core` branch to make this work. 

## Summary by CodeRabbit

- **New Features**
  - Updated user management functionality, including streamlined user creation and trade index updates.
  - Enhanced user rating process with direct database interactions.
  - Modified primary key in the users table to improve data integrity.
  - Added new asynchronous functions for user management in the database.
  - Introduced new queries for updating user ratings and trade indexes in the users table.
  - Added seller and buyer identity fields in order objects to enhance tracking.

- **Bug Fixes**
  - Improved error handling for user creation and rating updates.

- **Documentation**
  - Updated comments and formatting in the configuration file for clarity.

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -96,14 +99,14 @@ async fn check_trade_index(pool: &Pool<Sqlite>, event: &UnwrappedGift, msg: &Mes
...
                         pubkey: event.sender.to_string(),
                         last_trade_index,
                         ..Default::default()
                     };
-                    if new_user.create(pool).await.is_ok() {
-                        tracing::info!("Added new user for rate");
-                    }
+                    add_new_user(pool, new_user.pubkey, new_user.last_trade_index)
+                        .await
+                        .unwrap();
```
[the remark this answers is not in the document]
Comment (arkanoider) [reply src/app.rs:109]: Done.

File: src/app/rate_user.rs [hunk 2]
```diff
@@ -132,39 +133,53 @@ pub async fn update_user_reputation_action(
...
+    let old_rating = user_to_vote.total_rating;
+    let last_rating = user_to_vote.last_rating;
+    let new_rating = old_rating + (last_rating - old_rating) / (user_to_vote.total_reviews);
+
+    user_to_vote.last_rating = rating.into();
+    user_to_vote.total_reviews += 1;
+
+    // Assign new total rating to review
+    user_to_vote.total_rating = new_rating;
+
```
[the remark this answers is not in the document]
Comment (arkanoider) [reply src/app/rate_user.rs:150]: mmmmhhh...@grunch is the rabbit right here??

File: src/db.rs [hunk 3]
```diff
@@ -327,6 +327,72 @@ pub async fn is_user_present(pool: &SqlitePool, public_key: String) -> anyhow::R
...
+        "#,
+    )
+    .bind(public_key)
+    .bind(last_trade_index)
+    .bind(created_at.to_string())
+    .fetch_one(pool)
+    .await?;
+
+    Ok(user)
+}
```
[the remark this answers is not in the document]
Comment (arkanoider) [reply src/db.rs:350]: Done!

File: src/db.rs [hunk 4]
```diff
@@ -327,6 +327,72 @@ pub async fn is_user_present(pool: &SqlitePool, public_key: String) -> anyhow::R
...
+            UPDATE users SET trade_index = ?1 WHERE pubkey = ?2
+        "#,
+    )
+    .bind(trade_index)
+    .bind(public_key)
+    .fetch_one(pool)
+    .await?;
+
+    Ok(user)
+}
```
[the remark this answers is not in the document]
Comment (arkanoider) [reply src/db.rs:368]: Done!

File: src/db.rs [hunk 5]
```diff
@@ -327,6 +327,72 @@ pub async fn is_user_present(pool: &SqlitePool, public_key: String) -> anyhow::R
...
+    .bind(min_rating)
+    .bind(max_rating)
+    .bind(total_reviews)
+    .bind(total_rating)
+    .bind(public_key)
+    .fetch_one(pool)
+    .await?;
+
+    Ok(user)
+}
```
[the remark this answers is not in the document]
Comment (arkanoider) [reply src/db.rs:394]: Done!

File: src/app/rate_user.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (grunch) [reply src/app/rate_user.rs:150]: 🐰 is right, `total_reviews` should be incremented before calculation

[the remark this answers is not in the document]
Comment (arkanoider) [reply src/app/rate_user.rs:150]: > 🐰 is right, `total_reviews` should be incremented before calculation

Will fix as my master rabbit said! @coderabbitai I won't permit to have doubts on you! Pardon me!

[the remark this answers is not in the document]
Comment (arkanoider) [reply src/app/rate_user.rs:150]: Done!

File: src/app/rate_user.rs [hunk 6]
```diff
@@ -132,39 +133,55 @@ pub async fn update_user_reputation_action(
...
+    // Update user reputation
+    // Going on with calculation
+    let old_rating = user_to_vote.total_rating;
+    let last_rating = user_to_vote.last_rating;
+    
+    // increment first
+    user_to_vote.total_reviews += 1;
+    // recompute rating
+    let new_rating = old_rating + (rating.into() - old_rating) / (user_to_vote.total_reviews as f64);
```
Comment (grunch) [inline src/app/rate_user.rs:147]: please check the formula isn't right, you are now calculating with the last rating instead of just save it in db

Comment (arkanoider) [reply src/app/rate_user.rs]: Too many AI fighting each other. 

Now should be ok:

```Rust
  // Get counter to vote from db
    let mut user_to_vote = is_user_present(pool, counterpart.clone()).await?;

    // Update user reputation
    // Going on with calculation
    let old_rating = user_to_vote.total_rating;
    // increment first
    user_to_vote.total_reviews += 1;
    // recompute rating
    let new_rating = old_rating + (rating as i64 - old_rating) / (user_to_vote.total_reviews);

    user_to_vote.last_rating = rating.into();

    // Assign new total rating to review
    user_to_vote.total_rating = new_rating;
    // Create new rating event
    let reputation_event = Rating::new(
        user_to_vote.total_reviews as u64,
        user_to_vote.total_rating as f64,
        rating as u8,
        user_to_vote.min_rating as u8,
        user_to_vote.max_rating as u8,
    )
    .to_tags()?;

    // Save new rating to db
    if let Err(e) = update_user_rating(
        pool,
        user_to_vote.pubkey,
        user_to_vote.last_rating,
        user_to_vote.min_rating,
        user_to_vote.max_rating,
        user_to_vote.total_reviews,
        user_to_vote.total_rating,
    )
    .await
    {
        return Err(Error::msg(format!("Error updating user rating : {}", e)));
    }

File: src/app/rate_user.rs [hunk 7]
```diff
@@ -90,10 +91,10 @@ pub async fn update_user_reputation_action(
 
     // Find the counterpart public key
     if message_sender == buyer {
-        counterpart = seller;
+        counterpart = order.master_seller_pubkey.unwrap();
         buyer_rating = true;
     } else if message_sender == seller {
-        counterpart = buyer;
+        counterpart = order.master_buyer_pubkey.unwrap();
         seller_rating = true;
```
[the remark this answers is not in the document]
Comment (arkanoider) [reply src/app/rate_user.rs:98]: Done

**Review by Catrya — COMMENTED** [review]
Hi guys, great work!

In the 38383 events  it always appears `"rating", "none"` , but in the database there are rates. I'm sure you've already seen that.

Minor things:
- If a user tries to rate himself with a pending order of his (I don't think many try it), this appears in the logs:
 ```
2024-12-24T01:11:16.230136Z WARN mostrod::app: Error in RateUser with context Missing buyer pubkey // if is a purchase order:  Missing seller pubkey
```
Maybe it should return CantDo

I tried to break the rate and mostrod but I couldn't :) for now everything else I saw is from mostro-cli and I left it in the review of the rate-user PR in mostro-cli

**Comment by arkanoider** [comment]
> In the 38383 events it always appears `"rating", "none"` , but in the database there are rates. I'm sure you've already seen that.

Did not see! Thanks @Catrya will fix that!

**Review by Catrya — COMMENTED** [review]
Hi guys, I found some other things:

- In the 38383 rate event when an order is completed, in the `d` tag, the users' i0 pubkey is being published.
I think that if reputation is not going to be left as something global, but rather each user has a different reputation in each mostro, then there is no need to leave that event, even if it were published in the `d` tag of the trade key, what do you think?

- After restarting the mostro database, I get this in the logs
```
2024-12-24T20:08:05.254586Z ERROR mostrod::app: Error creating new user: no column found for name: is_admin
```

**Comment by arkanoider** [comment]
> In the 38383 rate event when an order is completed, in the d tag, the users' i0 pubkey is being published.\nI think that if reputation is not going to be left as something global, but rather each user has a different reputation in each mostro, then there is no need to leave that event, even if it were published in the d tag of the trade key, what do you think?

First of all... why working on Xmas eve? 🤣
Jokes apart, good pick will check you're right and I think that also @grunch (and mee too) want to just save all on local db in the next steps.

**Comment by arkanoider** [comment]
> After restarting the mostro database, I get this in the logs

I observed this too today. Entry in db seems ok but there is that error raising when a new user is saved the first time.  Will delve in this after Xmas!

**Comment by Catrya** [comment]
> First of all... why working on Xmas eve? 🤣
> Jokes apart, good pick will check you're right and I think that also @grunch (and mee too) want to just save all on local db in the next steps. 

haha look who's talking! 😂In Cuba we don't celebrate Christmas like in other countries. Anyway, I'm done for today💆‍♀️

**Comment by grunch** [comment]
> > In the 38383 rate event when an order is completed, in the d tag, the users' i0 pubkey is being published.\nI think that if reputation is not going to be left as something global, but rather each user has a different reputation in each mostro, then there is no need to leave that event, even if it were published in the d tag of the trade key, what do you think?
> 
> First of all... why working on Xmas eve? 🤣 Jokes apart, good pick will check you're right and I think that also @grunch (and mee too) want to just save all on local db in the next steps.

please let's keep posting the rate on nostr for now, as we were doing before 😃

**Comment by grunch** [comment]
> Hi guys, great work!
> 
> In the 38383 events it always appears `"rating", "none"` , but in the database there are rates. I'm sure you've already seen that.
> 
> Minor things:
> 
>     * If a user tries to rate himself with a pending order of his (I don't think many try it), this appears in the logs:
> 
> 
> ```
> 2024-12-24T01:11:16.230136Z WARN mostrod::app: Error in RateUser with context Missing buyer pubkey // if is a purchase order:  Missing seller pubkey
> ```
> 
> Maybe it should return CantDo
> 
> I tried to break the rate and mostrod but I couldn't :) for now everything else I saw is from mostro-cli and I left it in the review of the rate-user PR in mostro-cli

It's ok to show messages like that in logs, CantDo is used on messages not in logs

**Review by grunch — APPROVED** [review]
tACK

great work, let's merge this 👶🏽

**Comment by Catrya** [comment]
> It's ok to show messages like that in logs, CantDo is used on messages not in logs


I meant to send the message CantDo to the user, and in the logs leave it as it is. But in any case a user who tries to rate a self order in status pending knows that he cannot do it, so it is not so important to mostrod respond with CantDo

**Pull request merged**